### PR TITLE
[Coord along edge] Find coordinate for a point along edge

### PR DESCRIPTION
**Pull request by MathiasMehl**
Given an edge, and an int n indicating number of meters from start of the edge, position_on_edge() will return the coordinate for n meters from the start of the edge, following the shapes of the edge.

### CODE REVIEW

File: utility.py [hunk 1]
```diff
@@ -35,4 +35,31 @@ def distance(pos1: Tuple[float, float], pos2: Tuple[float, float]):
...
+
+
+def position_on_edge(edge: sumolib.net.edge.Edge, pos: int):
+    """
+    :return: coordinate for pos meters from the start of the edge, following any shapes along edge
+    """
+    # Go through pair of coords, until meeting an edge, where if we travel through it, we have moved more than pos
+    # meters in total
+    remaining_distance = pos
+    for coord1, coord2 in (edge.getShape()[i:i + 2] for i in range(0, int(len(edge.getShape()) / 2 * 2), 2)):
```
Comment (falkecarlsen) [inline utility.py]:
```suggestion
    for coord1, coord2 in (edge.getShape()[i:i + 2] for i in range(0, int(len(edge.getShape())), 2)):
```
That seems a bit unnecessary

**Review by falkecarlsen — APPROVED** [review]

File: utility.py (hunk 1 above)
Comment (MathiasMehl) [reply utility.py]: I have no idea what you are talking about

Comment (NicEastvillage) [inline utility.py]: The division and mulitplication in `int(...) / 2 * 2` cancels out

File: utility.py [hunk 2]
```diff
@@ -35,4 +35,31 @@ def distance(pos1: Tuple[float, float], pos2: Tuple[float, float]):
...
+
+    # Normalize it by dividing by its own length
+    unit_vec = vec / np.linalg.norm(vec)
+
+    # Scale by remaining distance
+    unit_vec_scaled = unit_vec * remaining_distance
+
+    # Add this scaled vector to the start point, to find the correct coord that is at remaining distance from this
+    # coord, to coord2
+    return coord1[0] + unit_vec_scaled[0], coord1[1] + unit_vec_scaled[1]
```
Comment (NicEastvillage) [inline utility.py:65]: What the heck, is this the coords from the for-loop? Are they still in scope??

**Review by NicEastvillage — CHANGES_REQUESTED** [review]
Nice! Just one (two?) things.
I will make city gates use this as well asap

File: utility.py (hunk 1 above)
Comment (falkecarlsen) [reply utility.py]: Seems fixed in 796b834 :+1:

Comment (falkecarlsen) [reply utility.py]: 796b834

File: utility.py (hunk 2 above)
Comment (MathiasMehl) [reply utility.py:65]: Yes.
image

Comment (NicEastvillage) [reply utility.py:65]: billede

python why

**Review by NicEastvillage — APPROVED** [review]

**Pull request merged**